### ISSUE TITLE
MSK AWS Terraform cluster Issue

### ISSUE THREAD

**Issue by navinsnn53** [terraform-providers/terraform-provider-aws#8793]
Hi , I have tried creating AWS MSK cluster with reference from the document shared. But I am getting the error 

"Error: error creating MSK cluster: BadRequestException: The parameter value contains one or more characters that are not valid.
status code: 400, request id: e2589e6a-8161-11e9-8f31-6f8877605e30". 

Below is the terraform code used. Please help me to fix the error.

resource "aws_msk_cluster" "msk_cluster" {
cluster_name = "Testing_Cluster"
kafka_version = "2.1.0"
number_of_broker_nodes = 3

broker_node_group_info {
instance_type = "kafka.m5.large"
client_subnets = [
"${aws_subnet.subnet_a.id}",
"${aws_subnet.subnet_b.id}",
"${aws_subnet.subnet_c.id}",
]
ebs_volume_size = 5
security_groups = [ "${aws_security_group.MSK_Sg.id}" ]
}

tags = {
Name = "Cluster_MSK"
}
}

**Comment by bflad** [comment]
My hunch would be that the MSK service does not allow the usage of underscores (`_`) in cluster names. Can you try removing the underscore and seeing if that resolves the issue?

The [MSK API Reference](https://docs.aws.amazon.com/msk/1.0/apireference/clusters.html#clusters-prop-createclusterrequest-clustername), does not provide too much guidance for string validation on fields such as the cluster name. We currently only implement the known string length check within the Terraform resource for the `cluster_name` argument: https://github.com/terraform-providers/terraform-provider-aws/blob/5619c0e0ac94f4dd645dd2f388c19795b5800341/aws/resource_aws_msk_cluster.go#L83

**Comment by navinsnn53** [comment]
@bflad  That worked.... Thank you soo much.

**Comment by navinsnn53** [comment]
I have remodified the code with "  count " argument so that I will be able to associate multiple subnets to a single Route table. When using az_distribution with count of not able create MSK cluster. Below are the code for reference 

main.tf
#### MSK CLUSTER CREATION
resource "aws_msk_cluster" "mskcluster" {
	  cluster_name           = "TestingCluster"
	  kafka_version          = "2.1.0"
	  number_of_broker_nodes = 3
	  count = "${length(var.subnet_ip_range)}"
	  broker_node_group_info {
	    instance_type  = "kafka.m5.large"
	    client_subnets = [
	  		"${element(aws_subnet.subnet.*.id, count.index)}"
	    ]
	    ebs_volume_size = 5
	    security_groups = [ "${aws_security_group.MSK_Sg.id}" ]
	    az_distribution = ["${var.availability_zones[count.index]}"]
	  }
	  tags = {
	    Name = "Cluster_MSK"
	  }
	}

#### Create Subnet in AZ
	resource "aws_subnet" "subnet" {
	  count = "${length(var.subnet_ip_range)}"
	  vpc_id     = "${aws_vpc.MSK_vpc.id}"     
	  cidr_block = "${var.subnet_ip_range[count.index]}"
	  availability_zone = "${var.availability_zones[count.index]}"
	  map_public_ip_on_launch = true

	  tags = {
	    Name = "${var.name[count.index]}"
	  }
	}

variable.tf

# Subnet IP DETAILS

variable "subnet_ip_range" {
	default = ["10.123.4.0/24","10.123.5.0/24","10.123.6.0/24"]
	type = "list"
}

# Subnet availability_zones DETAILS

variable "availability_zones" {
	default = ["us-east-1a", "us-east-1b","us-east-1c"]
}

#Subnet Tag Names

variable "name" {
	default = ["Subnet_A","Subnet_B","Subnet_C"]
}

Error " 152: 	    az_distribution = ["${var.availability_zones[count.index]}"]

Inappropriate value for attribute "az_distribution": string required."

**Comment by bflad** [comment]
The [`az_distribution` argument](https://www.terraform.io/docs/providers/aws/r/msk_cluster.html#az_distribution) maps to  [`BrokerAzDistribution` in the API](https://docs.aws.amazon.com/msk/1.0/apireference/clusters.html#clusters-model-brokerazdistribution), which currently accepts only `DEFAULT` as a valid value. e.g.

```hcl
resource "aws_msk_cluster" "mskcluster" {
  # ... other configuration ...
  broker_node_group_info {
    az_distribution = "DEFAULT"
  }
}
```

If the resource is not able to successfully create without that optional configuration, it seems like a bug in the resource. I will label this issue for future tracking. For other issues please open a new GitHub issue so it can be separately tracked (issues with multiple items to complete tend to cause problems).

By the way, you can use [triple backtick code blocks in GitHub comments](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) to format code. The above markdown is:

``````
```hcl
resource "aws_msk_cluster" "mskcluster" {
  # ... other configuration ...
  broker_node_group_info {
    az_distribution = "DEFAULT"
  }
}
```
``````

**Comment by bflad** [comment]
It appears that our acceptance test `TestAccAWSMskCluster_basic` already covers the case where `az_distribution` is omitted from the configuration, so there is not an issue with not providing that argument in the configuration. We should document its only acceptable value though. 👍 

I believe this error you are seeing is due to how you are using `count`:

```
Error: "error creating MSK cluster: BadRequestException: The number of Availability Zones in the BrokerAZDistribution parameter must be equal to the number of client subnets.
```

What are you attempting to create? If it is:
* Single Kafka cluster with three nodes: Do not use `count` and set `client_subnets = ["${aws_subnet.subnet.*.id}"]` (so you are passing the three subnet IDs to this argument)
* Three Kafka clusters with three nodes each: Set `client_subnets = ["${aws_subnet.subnet.*.id}"]` (so you are passing the three subnet IDs to this argument)

**Comment by navinsnn53** [comment]
Hi @bflad I was able to create without using count. But was trying to create using count and passing subnets. I did try the above solution but it was throwing an error. I am trying to create single cluster with three nodes and trying to pass subnets to 'broker_node_group_info'. Below is the error 

"Error: Incorrect attribute value type on main.tf line 144, in resource "aws_msk_cluster" "mskcluster":
 144:    client_subnets = ["${aws_subnet.subnet.*.id}"]
Inappropriate value for attribute "client_subnets": element 0: string
required."
